### PR TITLE
Break down scene time in FrameStats

### PR DESCRIPTION
**Pull request by markschlosseratbentley**
This PR breaks down the scene time in FrameStats into some potentially helpful categories.

### CODE REVIEW

File: core/frontend/src/Viewport.ts [hunk 1]
```diff
@@ -2209,7 +2216,9 @@ export abstract class Viewport implements IDisposable {
       this.setupFromView();
 
     if (this._selectionSetDirty) {
+      this._frameStatsCollector.beginTime("hiliteTime");
```
Comment (pmconne) [inline core/frontend/src/Viewport.ts:2219]: Not useful - it just assigns to a variable.

File: core/frontend/src/Viewport.ts [hunk 2]
```diff
@@ -2236,13 +2245,16 @@ export abstract class Viewport implements IDisposable {
     }
 
     if (overridesNeeded) {
+      this._frameStatsCollector.beginTime("featureSymbologyTime");
```
Comment (pmconne) [inline core/frontend/src/Viewport.ts:2248]: Not useful - just assigns to a variable.

File: core/frontend/src/Viewport.ts [hunk 3]
```diff
@@ -2258,40 +2270,49 @@ export abstract class Viewport implements IDisposable {
...
+      this._frameStatsCollector.beginTime("decorationsTime");
       const decorations = new Decorations();
       this.addDecorations(decorations);
       target.changeDecorations(decorations);
       this._decorationsValid = true;
       isRedrawNeeded = true;
+      this._frameStatsCollector.endTime("decorationsTime");
     }
 
+    this._frameStatsCollector.beginTime("flashTime");
```
Comment (pmconne) [inline core/frontend/src/Viewport.ts:2296]: Not useful.

File: core/frontend/src/Viewport.ts [hunk 4]
```diff
@@ -2258,40 +2270,49 @@ export abstract class Viewport implements IDisposable {
...
     let requestNextAnimation = false;
     if (this.processFlash()) {
       target.setFlashed(undefined !== this._flashedElem ? this._flashedElem : Id64.invalid, this.flashIntensity);
       isRedrawNeeded = true;
       requestNextAnimation = undefined !== this._flashedElem;
     }
+    this._frameStatsCollector.endTime("flashTime");
 
     target.onBeforeRender(this, (redraw: boolean) => {
```
Comment (pmconne) [inline core/frontend/src/Viewport.ts:2305]: Record this - this is one of the places Open Cities Planner hooks.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/frontend/src/Viewport.ts (hunk 4 above)
Comment (pmconne) [reply core/frontend/src/Viewport.ts:2305]: Can you record IModelFrameLifecycle.onRenderOpaque in SceneCompositor.ts? That's where OCP does their heaviest lifting.

Comment (markschlosseratbentley) [reply core/frontend/src/Viewport.ts:2305]: Yes, it should be simple to wrap `IModelFrameLifecycle.onRenderOpaque.raiseEvent` with timing tracking in SceneCompositor.ts.

What line were you originally referring to?  `target.onBeforeRender()` in Viewport.ts?

Comment (pmconne) [reply core/frontend/src/Viewport.ts:2305]: Comments appear directly below the line they reference - in this case line 2305 in Viewport.ts.

Comment (markschlosseratbentley) [reply core/frontend/src/Viewport.ts:2300]: Thanks for the feedback.  I have added 'onBeforeRenderTime' and 'onRenderOpaqueTime'.

File: core/frontend/src/Viewport.ts (hunk 3 above)
Comment (markschlosseratbentley) [reply core/frontend/src/Viewport.ts]: Removed.

File: core/frontend/src/Viewport.ts (hunk 2 above)
Comment (markschlosseratbentley) [reply core/frontend/src/Viewport.ts]: Removed.

File: core/frontend/src/Viewport.ts (hunk 1 above)
Comment (markschlosseratbentley) [reply core/frontend/src/Viewport.ts]: Removed.

File: core/frontend/src/render/webgl/Target.ts [hunk 5]
```diff
@@ -593,19 +593,17 @@ export abstract class Target extends RenderTarget implements RenderTargetDebugCo
     }
   }
 
-  private _frameStatsCollector = new FrameStatsCollector();
+  private _frameStatsCollector?: FrameStatsCollector;
```
Comment (pmconne) [inline core/frontend/src/render/webgl/Target.ts:596]: Just a thought, if you leave this like it was previously you won't have to check for `undefined` every time you access it.

**Review by pmconne — APPROVED** [review]

File: core/frontend/src/render/webgl/Target.ts (hunk 5 above)
Comment (markschlosseratbentley) [reply core/frontend/src/render/webgl/Target.ts:596]: You make a good point. I had considered that and for some reason went with this approach, but I am not a fan of the undefined checks... I'll see what I can do.

Comment (markschlosseratbentley) [reply core/frontend/src/render/webgl/Target.ts:596]: It is a bit more complicated now because both Viewport and Target need access to the collector.

Comment (pmconne) [reply core/frontend/src/render/webgl/Target.ts:596]: I'm saying: keep assignFrameStats and make it overwrite the one you initially created on the Target. Similar to how we default-initialize _hilites then overwrite it later.

Comment (markschlosseratbentley) [reply core/frontend/src/render/webgl/Target.ts:596]: Yep, we're saying the same thing. I agree - it's a better pattern. Will push shortly.

Comment (markschlosseratbentley) [reply core/frontend/src/render/webgl/Target.ts]: I pushed this change - can you mark 'resolved' if looks good?

**Comment by markschlosseratbentley** [comment]
@Mergifyio backport release/2.16.x

**Pull request merged**